src/datatable/DatatableDataModel.md

```diff
@@ -0,0 +1,45 @@
+# Data model for a Datatable
+
+- a requirements/specifications document so that we can decide on the best approach to developing a XML schema for Datatables
+
+- this document will either need to be updated to be consistent with the developed schema, or deleted
+
+- initially, for determining approach, we should focus on the MUSTs and worry less about the OPTIONALs
+
+- a Datatable is similar to database table, it has one or more **columns** (a.k.a _fields_) and for each column there is a value in each **row**
+
+- although there is an emphasis on a column wise schema for Datatables this does not necessarily mean that the in-memory representation needs to be column-wise 
+
+- a Datatable can have OPTIONAL meta data such as **name**, **title**, **description** and **authors** which the user SHOULD be able to add and modify
+
+- Each **column**,
+
+    - MUST have a **name**, this name SHOULD be able to be set and changed by the user, and/or be defined by the external data source (e.g. a CSV file with a header)
+
+    - MUST have a **type** which specifies the _type_ of data that can be in that column:
+
+        - the user SHOULD be able to change the type of a column
+
+        - types might include string, number, integer, boolean, date
+
+        - the default type is "string"
+
+        - having types allows for editing restrictions on the type of data that a user can enter into a cell
+
+        - having types allows for optimisations in storage and analysis across the Datatable
+
+    - can have one or more OPTIONAL **constraints** which specifies the acceptable _values_ in each column
+
+    - can have an OPTIONAL **title** string which the user SHOULD be able to modify
+
+    - can have an OPTIONAL **description** string which the user SHOULD be able to modify
+
+- Similarity and difference to `Sheets`:
+
+  - Superficially the UI for `Datatables` is similar to that for `Sheets` in that they are both grid based
+
+  - However, the _data model_ is quite different. Sheets can be though of as a graph of cells, each specifying their dependency on other cells via cell formulae. In a spreadsheet those cells are displayed in a grid. But that is really a UI convention, not a reflection of the underlying data model. The position of a Sheet cell in that sheet does not determine the data content of that cell. A cell in column A1 is not restricted to having the same type of value as a cell in A2 even though they are in the same column. In contrast, in a Datatable, cells in a column are restricted to having the same type.
```

Comment:
I think the key of success of the Sheet paradigm was the ability to mix data and code, with focus on the data side (as opposed to the flow/dependency part). It is not just a convention. It is a projection of the multi-dimension nature of a program: code, values, and dependencies, into one view.
In imperative programming you see the local dependencies, or order, but need a debugger to reveal the values.
In graphical programming you see the data-flow, plus a symbolic version of the code, not the values.